update ipv6 documentation and add dual-stack migration part (#1014)
gardener/gardener-extension-provider-gcp: docs/usage/ipv6.md

> # DualStack Support for Gardener GCP Extension

Reviewer: Not a native speaker, but in the AWS provider documentation we used `dual-stack` a lot instead of `DualStack`. We should probably be consistent.

@n-boshnakov Can you provide guidance in this regard?

Reviewer 2: As the more commonly used way of writing appears to be `dual-stack`, I'd suggest we stay consistent with the AWS documentation. This screenshot provides a good example on using the term in different cases - as a title, in the beginning of a sentence and within a sentence.

![Screenshot 2025-03-21 165359](https://github.com/user-attachments/assets/4e914272-9add-4244-8b5d-ac528fa3e651)

This document provides an overview of DualStack support for the Gardener GCP extension.
Furthermore it clarifies which components are utilized, how the infrastructure is setup and how a DualStack cluster can be provisioned.

> ---

Reviewer: Might be personal preference, but I can live without the multiple separators in this document. Therefore, I propose to drop all of them, e.g.

```suggestion
```

## Overview

> DualStack support allows Gardener GCP shoot clusters to leverage both IPv4 and IPv6 addresses.

Reviewer: ```suggestion
Gardener allows to create dual-stack shoot clusters on GCP. In this mode, both IPv4 and IPv6 are supported within the shoot cluster.
```

> ### Key Components

Reviewer: ```suggestion
### Key Components for Dual-Stack Support
```

> - **DualStack Subnets**: Separate subnets are created for nodes and services, with explicit IPv4 and external IPv6 ranges.

Reviewer: This makes it sound like there is an IPv4 node subnet and an additional IPv6 node subnet, which is not the case as far as I know. There is a special (additional) IPv6 subnet for services, but I would not position it as a key component, i.e. mention it in the next section only.

- **Ingress-GCE Component**: Responsible for creating DualStack (IPv4,IPv6) Load Balancers.

> - **Cloud Allocator for IPAM**: Manages the assignment of IPv4 and IPv6 ranges to nodes and pods.

Reviewer: Let's add links to the sections.

# Dual-stack Support for Gardener GCP Extension

This document provides an overview of dual-stack support for the Gardener GCP extension.
Furthermore it clarifies which components are utilized, how the infrastructure is setup and how a dual-stack cluster can be provisioned.

## Overview

Gardener allows to create dual-stack clusters on GCP. In this mode, both IPv4 and IPv6 are supported within the cluster.
This significantly expands the available address space, enables seamless communication across both IPv4 and IPv6 environments and ensures compliance with modern networking standards.

### Key Components for Dual-Stack Support

> - **[Dual-stack Subnets](#dual-stack-subnets)**: Separate subnets are created for nodes and services, with explicit IPv4 and external IPv6 ranges.

Reviewer 2: ```suggestion
- **[Dual-Stack Subnets](#dual-stack-subnets)**: Separate subnets are created for nodes and services, with explicit IPv4 and external IPv6 ranges.
```

> - This subnet is dedicated to IPv6 services. It is created due to GCP's limitation of not supporting IPv6 reservation for services.

Reviewer: ```suggestion
- This subnet is dedicated to IPv6 services. It is created due to GCP's limitation of not supporting IPv6 range reservations.
```

### 3. **Internal Subnet (optional)**

> - This subnet is dedicated for internal load balancer. Currently are only internal IPv4 loadbalancer supported. They are provisioned by the Cloud Controller Manager (CCM).

Reviewer: ```suggestion
- This subnet is dedicated for internal load balancer. Currently, only internal IPv4 loadbalancer are supported. They are provisioned by the Cloud Controller Manager (CCM).
```

---

## Additional Components

> ### 1. **Ingress-GCE**

Reviewer: Let's combine the headers. There is no need for the additional structure and/or a numbering with just one entity.

Reviewer: Maybe it makes even more sense to move the content to the Load Balancer section.

- The ingress-gce is a mandatory component for DualStack clusters. It is responsible for creating DualStack (IPv4,IPv6) Load Balancers. This is necessary because the GCP Cloud Controller Manager does not support provisioning IPv6 Load Balancer.

---

> ## Cloud Allocator (IPAM)

Reviewer: Let's drop this whole section. It is not relevant for users as pod IP address management should just work.

If you want to keep it move it to a different document for extension developers.

## Creating a DualStack Cluster

To create a DualStack cluster, both IP families (IPv4,IPv6) need to be specified under `spec.networking.ipFamilies`. Below is an example of a DualStack shoot cluster configuration:

```yaml

> apiVersion: core.gardener.cloud/v1beta1

Reviewer: Please remove the annotations from the example.

  ...
```

## Migration of IPv4-Only Shoot Clusters to Dual-Stack

To trigger a migration of an IPv4 cluster to dual-stack, IPv6 needs to be added to `spec.networking.ipFamilies` in the shoot specification.

> , a constraint of type `DualStackNodesMigrationReady` is added to the shoot status. It is in state `False` until all nodes have an IPv4 and IPv6 address assigned.

Reviewer 3: This sentence starts with ",". I assume this is a typo?

Changing the `ipFamilies` field triggers immediately an infrastructure reconciliation, where the infrastructure is reconfigured to additionally support IPv6. During this infrastructure migration process get the subnets an external IPv6 range and the node subnet gets a secondary IPv4 range. Pod specific cloud routes are deleted from the VPC route table and alias IP ranges for the pod routes are added to the NIC of kubernetes nodes/instances.

> With the next node roll-out which is a manual step and will not be triggered automatically, all nodes will get an IPv6 address and an IPv6 prefix for pods assigned. When all nodes have IPv4 and IPv6 pod ranges, will the status of the `DualStackNodesMigrationReady` constraint be changed to `True`.

Reviewer 2: ```suggestion
With the next node roll-out, which is a manual step and will not be triggered automatically, all nodes will get an IPv6 address and an IPv6 prefix for pods assigned. When all nodes have IPv4 and IPv6 pod ranges, the status of the `DualStackNodesMigrationReady` constraint will be changed to `True`.
```

Once all nodes are migrated, the remaining control plane components and the Container Network Interface (CNI) are configured for dual-stack networking and the migration constraint is removed at the end of this step.

## Load Balancer Service Configuration

> To create a dual-stack LoadBalancer the `spec.ipFamilies` and `spec.ipFamilyPolicy` field needs to be specified in the kubernetes service.

Reviewer 2: ```suggestion
To create a dual-stack LoadBalancer, the `spec.ipFamilies` and `spec.ipFamilyPolicy` field needs to be specified in the Kubernetes service.
```

  ...
```

---

## Migration of IPv4-only Shoot Clusters to DualStack

> To trigger a migration of an IPv4 shoot cluster to DualStack, both ipFamilies needs to be specified under `spec.networking.ipFamilies` in the shoot.

Reviewer: ```suggestion
To trigger a migration of an IPv4 shoot cluster to DualStack, IPv6 needs to be added to `spec.networking.ipFamilies` in the shoot specification.
```

Once the migration is triggered a constraint of type `ToDualStackMigration` is added to the shoot status. It is in progressing state.
With the next shoot maintenance, the infrastructure is migrated to IPv6. The subnets get an external IPv6 range, the node subnet gets the secondary IPv4 range. The pod specific cloud routes are deleted from the VPC route table and alias IP ranges for the pod routes are added to the NIC of kubernetes nodes/instances. After that the status of the constraint will be changed to `DualStackInfraReady`. 
With the next node roll-out, nodes will get IPv6 addresses and an IPv6 prefix for pods. When all nodes have IPv4 and IPv6 pod ranges, the status will be changed to `DualStackNodesReady`.

> The next reconcile will change all the remaining components to DualStack. Once the migration has finished the constraint will be removed with the next reconcile.

Reviewer: Might need changes after updates of the migration PR in gardener/gardener, e.g. the name of the constraint might change.

---

## Load Balancer Service Configuration

To create a DualStack LoadBalancer the `spec.ipFamilies` and `spec.ipFamilyPolicy` field needs to be specified in the kubernetes service.
An example configuration is shown below:

```
apiVersion: v1
kind: Service
metadata:
  name: nginx
  namespace: default
  annotations:
    cloud.google.com/l4-rbs: enabled
spec:
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - port: 12345
    protocol: TCP
    targetPort: 80
  selector:
    run: nginx
  type: LoadBalancer
```

> The required `cloud.google.com/l4-rbs: enabled` annotation for ingress gce is added automatically via webhook for services of `type: LoadBalancer`.

Reviewer: ```suggestion
The required annotation `cloud.google.com/l4-rbs: enabled` for ingress-gce is added automatically via webhook for services of `type: LoadBalancer`.
```

### Explanation
- **`cloud.google.com/l4-rbs: enabled`**: Ensures that `ingress-gce` properly reconciles the LoadBalancer service with IPv6 support.

> - **`ipFamilyPolicy` and `ipFamilies`**: Specify the DualStack configuration.

Reviewer: Not sure about the usefulness of this section. I would probably rather remove it.
WDYT?

### Internal Load Balancer

> - Internal IPv6 LoadBalancers are **not supported**.

Reviewer: ```suggestion
- Internal IPv6 LoadBalancers are currently **not supported**.
```

- To create internal IPv4 LoadBalancers, you can set one of the the following annotations:
  - `"networking.gke.io/load-balancer-type=Internal"`
  - `"cloud.google.com/load-balancer-type=internal"` (deprecated).

>   Internal load balancers are created by cloud-controller-manger and get an an IPv4 address from the internal subnet.

Reviewer: ```suggestion

Internal load balancers are created by cloud-controller-manger and get an IPv4 address from the internal subnet.
```

> ## Key Benefits of DualStack Support

Reviewer: Let's remove this section and use the important parts of the content in the introduction/overview sections.

---

> ## Summary

Reviewer: I see no need for this section as well. If we want to keep parts of the content, let's move it to the other sections.
